lib/spack/docs/build_systems/makefilepackage.rst

```diff
@@ -0,0 +1,295 @@
+.. _makefilepackage:
+
+---------------
+MakefilePackage
+---------------
+
+The most primitive build system a package can use is a plain Makefile.
+Makefiles are simple to write for small projects, but they usually
+require you to edit the Makefile to set platform and compiler-specific
+variables.
+
+^^^^^^
+Phases
+^^^^^^
+
+``MakefilePackage`` comes with 3 phases:
+
+#. ``edit`` - edit the Makefile
+#. ``build`` - build the project
+#. ``install`` - install the project
+
+By default, ``edit`` does nothing, but you can override it to replace
+hard-coded Makefile variables. The ``build`` and ``install`` phases
+run:
+
+.. code-block:: console
+
+   $ make
+   $ make install
+
+
+^^^^^^^^^^^^^^^
+Important files
+^^^^^^^^^^^^^^^
+
+The main file that matters for a ``MakefilePackage`` is the Makefile.
+This file will be named one of the following ways:
+
+* GNUmakefile (only works with GNU Make)
+* Makefile (most common)
+* makefile
+
+``Makefile`` is the most common name.
+
+Some Makefiles also *include* other configuration files. Check for an
+``include`` directive in the Makefile.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Build system dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Spack assumes that the operating system will have a valid ``make`` utility
+installed already, so you don't need to add a dependency on ``make``.
+However, if the package uses a ``GNUmakefile`` or the developers recommend
+using GNU Make, you should add a dependency on ``gmake``:
+
+.. code-block:: python
+
+   depends_on('gmake', type='build')
+
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Types of Makefile packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Most of the work involved in packaging software that uses Makefiles
+involves overriding or replacing hard-coded variables. Many packages
+make the mistake of hard-coding compilers, usually for GCC or Intel.
+This is fine if you happen to be using that particular compiler, but
+Spack is designed to work with *any* compiler, and you need to ensure
+that this is the case.
+
+Depending on how the Makefile is designed, there are 4 common strategies
+that can be used to set or override the appropriate variables:
+
+"""""""""""""""""""""
+Environment variables
+"""""""""""""""""""""
+
+Make has multiple types of assignment operators. Some Makefiles
+use ``=`` to assign variables. The only way to override these
+variables is to edit the Makefile or override them on the command-line.
+However, Makefiles that use ``?=`` for assignment honor environment
+variables. Since Spack already sets ``CC``, ``CXX``, ``F77``, and ``FC``,
+you won't need to worry about setting these variables. If there are
+any other variables you need to set, you can do this in the ``edit``
+method:
+
+.. code-block:: python
+
+   def edit(self, spec, prefix):
+       env['PREFIX'] = prefix
```

Comment:
Where does the `env` variable come from?  Why is this different from `setup_environment()`?

Reply:
> Where does the `env` variable come from?

https://github.com/spack/spack/blob/develop/lib/spack/spack/build_environment.py#L364

> Why is this different from `setup_environment()`?

The only real difference I've seen is that `setup_environment` is run before `edit`, and before we switch to the build directory. So things that need to know the current build directory will fail unless you set them in `edit`.

Reply:
This information needs to be included in the docs.